wideq/core.py

```diff
@@ -240,166 +345,218 @@ def parse_oauth_callback(url):
-    # https://us.lgeapi.com/datetime
-    # But we can also just generate a timestamp.
-    timestamp = datetime.datetime.utcnow().strftime(DATE_FORMAT)
+    return out["access_token"], out["refresh_token"]
 
-    # The signature for the requests is on a string consisting of two
-    # parts: (1) a fake request URL containing the refresh token, and (2)
-    # the timestamp.
-    req_url = ('/oauth2/token?grant_type=refresh_token&refresh_token=' +
-               refresh_token)
-    sig = oauth2_signature('{}\n{}'.format(req_url, timestamp),
-                           OAUTH_SECRET_KEY)
 
-    headers = {
-        'lgemp-x-app-key': OAUTH_CLIENT_KEY,
-        'lgemp-x-signature': sig,
-        'lgemp-x-date': timestamp,
-        'Accept': 'application/json',
-    }
+def refresh_auth(oauth_root, refresh_token):
+    """Get a new access_token using a refresh_token.
 
-    with retry_session() as session:
-        res = session.post(token_url, data=data, headers=headers)
-    res_data = res.json()
+    May raise a `TokenError`.
+    """
+    out = auth_request(
+        oauth_root, {"grant_type": "refresh_token", "refresh_token": refresh_token}
+    )
 
-    if res_data['status'] != 1:
-        raise TokenError()
-    return res_data['access_token']
+    return out["access_token"]
 
 
 class Gateway(object):
-    def __init__(self, auth_base, api_root, oauth_root, country, language):
+    def __init__(self, auth_base, api_root, api2_root, country, language):
         self.auth_base = auth_base
         self.api_root = api_root
-        self.oauth_root = oauth_root
+        self.api2_root = api2_root
         self.country = country
         self.language = language
 
     @classmethod
-    def discover(cls, country, language) -> 'Gateway':
-        """Load information about the hosts to use for API interaction.
+    def discover(cls, country, language):
+        gw = gateway_info(country, language)
+        return cls(gw["empUri"], gw["thinq1Uri"], gw["thinq2Uri"], country, language)
 
-        `country` and `language` are codes, like "US" and "en-US,"
-        respectively.
+    def oauth_url(self):
+        """Construct the URL for users to log in (in a browser) to start an
+        authenticated session.
         """
-        gw = lgedm_post(GATEWAY_URL,
-                        {'countryCode': country, 'langCode': language})
-        return cls(gw['empUri'], gw['thinqUri'], gw['oauthUri'],
-                   country, language)
 
-    def oauth_url(self):
-        return oauth_url(self.auth_base, self.country, self.language)
+        url = urljoin(self.auth_base, "spx/login/signIn")
+        query = urlencode(
+            {
+                "country": self.country,
+                "language": self.language,
+                "svc_list": SVC_CODE,
+                "client_id": CLIENT_ID,
+                "division": "ha",
+                "redirect_uri": OAUTH_REDIRECT_URI,
+                "state": uuid.uuid1().hex,
```

Comment:
As with the headers, is stuff like this really necessary? (Do we in fact need a `state` parameter?)

Reply:
Yes, I think we need, but not totally sure